### PR TITLE
Add a refresh operation for resources

### PR DESCRIPTION
**Pull request by shinyhappydan**
Part of #3680

### CODE REVIEW

File: delta/app/src/main/scala/ch/epfl/bluebrain/nexus/delta/routes/ResourcesRoutes.scala [hunk 1]
```diff
@@ -166,6 +166,20 @@ final class ResourcesRoutes(
                           )
                         }
                       },
+                      (pathPrefix("refresh") & put & pathEndOrSingleSlash) {
+                        operationName(s"$prefixSegment/resources/{org}/{project}/{schema}/{id}/source") {
```
Comment (shinyhappydan) [inline delta/app/src/main/scala/ch/epfl/bluebrain/nexus/delta/routes/ResourcesRoutes.scala:170]: We aren't taking revision as an argument. We will see later that revision is needed, so we take it from the record we load

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/ResourcesImpl.scala [hunk 2]
```diff
@@ -82,6 +82,22 @@ final class ResourcesImpl private (
...
+  override def refresh(
+      id: IdSegment,
+      projectRef: ProjectRef,
+      schemaOpt: Option[IdSegment]
+  )(implicit caller: Caller): IO[ResourceRejection, DataResource] = {
+    for {
+      projectContext        <- fetchContext.onModify(projectRef)
+      iri                   <- expandIri(id, projectContext)
+      schemeRefOpt          <- expandResourceRef(schemaOpt, projectContext)
+      (source, rev)         <- fetch(IdSegmentRef(iri), projectRef, schemaOpt).map(r => r.value.source -> r.rev)
```
Comment (shinyhappydan) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/ResourcesImpl.scala:94]: This is where we are getting the revision from. Not sure about this

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/model/ResourceCommand.scala [hunk 3]
```diff
@@ -107,6 +107,39 @@ object ResourceCommand {
...
+    *   the subject which created this event
+    */
+  final case class RefreshResource(
+      id: Iri,
+      project: ProjectRef,
+      schemaOpt: Option[ResourceRef],
+      source: Json,
+      compacted: CompactedJsonLd,
+      expanded: ExpandedJsonLd,
+      rev: Int,
```
Comment (shinyhappydan) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/model/ResourceCommand.scala:137]: Revision must be present here, `ResourceCommand` needs it

File: delta/app/src/main/scala/ch/epfl/bluebrain/nexus/delta/routes/ResourcesRoutes.scala (hunk 1 above)
Comment (imsdu) [reply delta/app/src/main/scala/ch/epfl/bluebrain/nexus/delta/routes/ResourcesRoutes.scala:170]: replace `/source` with `/refresh` in operation name

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala [hunk 4]
```diff
@@ -314,6 +336,28 @@ object Resources {
 
       }
 
+    def refresh(c: RefreshResource) = {
```
Comment (imsdu) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala:339]: It looks really like the one for updated, maybe we could reduce the boilerplate by introducing a common trait that `UpdateResource` and `RefreshResource` would implement ?

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/ResourcesImpl.scala (hunk 2 above)
Comment (imsdu) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/ResourcesImpl.scala:94]: I think we should use:
```scala
 log.stateOr(projectRef, iri, notFound)
```

**Review by imsdu — CHANGES_REQUESTED** [review]

File: delta/app/src/main/scala/ch/epfl/bluebrain/nexus/delta/routes/ResourcesRoutes.scala (hunk 1 above)
Comment (shinyhappydan) [reply delta/app/src/main/scala/ch/epfl/bluebrain/nexus/delta/routes/ResourcesRoutes.scala:170]: Not sure how this happened :)

File: docs/src/main/paradox/docs/delta/api/resources-api.md [hunk 5]
```diff
@@ -110,6 +110,27 @@ Response
 :   @@snip [updated.json](assets/resources/updated.json)
 
 
+## Refresh
+
+This operation refreshes the payload.
+
+This is equivalent of doing an update with the same source as the last revision of the resource. It is useful when the 
+schema or project contexts have changed, in order for the changes to be reflected in the resources
```
Comment (imsdu) [inline docs/src/main/paradox/docs/delta/api/resources-api.md:118]: At the end:
```
... resources.
```

File: docs/src/main/paradox/docs/delta/api/resources-api.md [hunk 6]
```diff
@@ -110,6 +110,27 @@ Response
 :   @@snip [updated.json](assets/resources/updated.json)
 
 
+## Refresh
+
+This operation refreshes the payload.
```
Comment (imsdu) [inline docs/src/main/paradox/docs/delta/api/resources-api.md:115]: `This operation refreshes the compacted and expanded representations of the resource.`

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala [hunk 7]
```diff
@@ -293,72 +316,104 @@ object Resources {
...
-        case Some(s)                                                       =>
-          // format: off
-          for {
-            (schemaRev, schemaProject) <- validate(s.project, c.schemaOpt.getOrElse(s.schema), c.caller, c.id, c.expanded)
-            types                       = c.expanded.cursor.getTypes.getOrElse(Set.empty)
-            time                       <- IOUtils.instant
-          } yield ResourceUpdated(c.id, c.project, schemaRev, schemaProject, types, c.source, c.compacted, c.expanded, s.rev + 1, time, c.subject)
-        // format: on
-
+        case Some(s)                                                       => IO.now(s)
```
Comment (imsdu) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala:327]: `IO.pure`

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala [hunk 8]
```diff
@@ -293,72 +316,104 @@ object Resources {
         case _ => IO.raiseError(ResourceAlreadyExists(c.id, c.project))
       }
 
-    def update(c: UpdateResource) =
+    def validateStateResourceExists(c: ModifyCommand) = {
```
Comment (imsdu) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala:319]: `resourceExists` ?

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala [hunk 9]
```diff
@@ -293,72 +316,104 @@ object Resources {
...
-          IO.raiseError(ResourceNotFound(c.id, c.project, c.schemaOpt))
-        case Some(s) if s.rev != c.rev                                     =>
-          IO.raiseError(IncorrectRev(c.rev, s.rev))
-        case Some(s) if c.schemaOpt.exists(cur => cur.iri != s.schema.iri) =>
-          IO.raiseError(UnexpectedResourceSchema(s.id, c.schemaOpt.get, s.schema))
-        case Some(s) if c.targetRev <= 0 || c.targetRev > s.rev            =>
-          IO.raiseError(RevisionNotFound(c.targetRev, s.rev))
-        case Some(s)                                                       =>
-          IOUtils.instant.map(ResourceTagAdded(c.id, c.project, s.types, c.targetRev, c.tag, s.rev + 1, _, c.subject))
+    def validateStateResourceCanBeModified(c: ModifyCommand) = {
```
Comment (imsdu) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala:331]: `resourceEditable` ?

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala [hunk 10]
```diff
@@ -293,72 +316,104 @@ object Resources {
...
+      validateStateResourceExists(c).flatMap { s =>
+        IO.raiseWhen(s.deprecated)(ResourceIsDeprecated(c.id)).as(s)
+      }
+    }
 
+    def validateStateTagExistsOnResource(c: ModifyCommand, tag: UserTag) = {
```
Comment (imsdu) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala:337]: As this one is not reused, what about including it directly in `deleteTag` ?
We would get something like:
```scala
for {
        s     <- resourceExists(c, c.tag)
        _      <- IO.raiseWhen(!s.tags.contains(tag))(TagNotFound(tag))
        time <- IOUtils.instant
      } yield ResourceTagDeleted(c.id, c.project, s.types, c.tag, s.rev + 1, time, c.subject)
```

**Review by imsdu — CHANGES_REQUESTED** [review]

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala (hunk 10 above)
Comment (shinyhappydan) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala:337]: To me it is more readable when the validation is both done in a single method, and named (the method name)

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala (hunk 8 above)
Comment (shinyhappydan) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala:319]: stateWhereResourceExists? As it is returning a state where that is true, or a reason why it isn't

**Comment by shinyhappydan** [comment]
image

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala (hunk 8 above)
Comment (imsdu) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/Resources.scala]: Ok for me :)

**Review by imsdu — APPROVED** [review]

**Pull request merged**